### PR TITLE
impl(generator): relocate RestConnection factory function to separate file

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10748)

### CODE REVIEW

File: generator/internal/connection_rest_generator.cc [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+      break;
+  }
+
+  auto result = HeaderOpenNamespaces();
+  if (!result.ok()) return result;
+  HeaderPrint(R"""(
+/**
+ * A factory function to construct an object of type `$connection_class_name$`
+ * that uses REST over HTTP as transport in lieu of gRPC. REST transport should
+ * only be used for services that do not support gRpc or if the existing network
```
Comment (dbolduc) [inline generator/internal/connection_rest_generator.cc:71]: s/gRpc/gRPC/

File: generator/internal/descriptor_utils.cc [hunk 2]
```diff
@@ -836,6 +837,12 @@ VarsDictionary CreateServiceVars(
   vars["connection_header_path"] =
       absl::StrCat(vars["product_path"],
                    ServiceNameToFilePath(descriptor.name()), "_connection.h");
+  vars["connection_rest_cc_path"] = absl::StrCat(
```
Comment (dbolduc) [inline generator/internal/descriptor_utils.cc:840]: can we have a unit test, such as:

https://github.com/googleapis/google-cloud-cpp/blob/8680bd69668a9e3b407c56cb103b79c1fa7f5157/generator/internal/descriptor_utils_test.cc#L239-L244

File: generator/internal/connection_rest_generator.cc [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    "// source: $proto_file_name$\n");
+  // clang-format on
+
+  // includes
+  CcPrint("\n");
+  CcLocalIncludes(
+      {vars("connection_rest_header_path"), vars("options_header_path"),
+       vars("connection_impl_rest_header_path"),
+       vars("option_defaults_header_path"),
+       vars("tracing_connection_header_path"),
```
Comment (dbolduc) [inline generator/internal/connection_rest_generator.cc:144]: I don't think this one is needed.

File: generator/internal/connection_rest_generator.cc [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    std::map<std::string, VarsDictionary> service_method_vars,
+    google::protobuf::compiler::GeneratorContext* context)
+    : ServiceCodeGenerator("connection_rest_header_path",
+                           "connection_rest_cc_path", service_descriptor,
+                           std::move(service_vars),
+                           std::move(service_method_vars), context) {}
+
+Status ConnectionRestGenerator::GenerateHeader() {
+  HeaderPrint(CopyrightLicenseFileHeader());
+  HeaderPrint(  // clang-format off
```
Comment (dbolduc) [inline generator/internal/connection_rest_generator.cc:39]: nit: can we copy a file that uses a raw string?

File: generator/internal/connection_rest_generator.cc [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+  HeaderCloseNamespaces();
+  // close header guard
+  HeaderPrint("\n#endif  // $header_include_guard$\n");
+  return {};
+}
+
+Status ConnectionRestGenerator::GenerateCc() {
+  CcPrint(CopyrightLicenseFileHeader());
+  CcPrint(  // clang-format off
```
Comment (dbolduc) [inline generator/internal/connection_rest_generator.cc:131]: same raw string comment.

File: generator/internal/descriptor_utils.cc [hunk 6]
```diff
@@ -836,6 +837,12 @@ VarsDictionary CreateServiceVars(
...
+      vars["product_path"], ServiceNameToFilePath(descriptor.name()),
+      "_rest_connection.cc");
+  vars["connection_rest_header_path"] = absl::StrCat(
+      vars["product_path"], ServiceNameToFilePath(descriptor.name()),
+      "_rest_connection.h");
```
Comment (dbolduc) [inline generator/internal/descriptor_utils.cc:845]: are we concerned that a customer might see the rest connection header, but have no idea how to enable it?

(unlikely, as we know that Spanner Admin APIs are not used.... but this problem seems avoidable)

File: generator/integration_tests/golden/v1/golden_thing_admin_rest_connection.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * - `google::cloud::golden_v1::GoldenThingAdminPolicyOptionList`
+ *
+ * @note Unexpected options will be ignored. To log unexpected options instead,
+ *     set `GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes` in the environment.
+ *
+ * @param options (optional) Configure the `GoldenThingAdminConnection` created by
+ * this function.
+ */
+std::shared_ptr<GoldenThingAdminConnection> MakeGoldenThingAdminConnectionRest(
+    Options options = {});
```
Comment (dbolduc) [inline generator/integration_tests/golden/v1/golden_thing_admin_rest_connection.h:57]: Can we have a unit test for this thing?

I know these are kinda weird to test... but can we do something like set a fake local endpoint, and make one client call we know will fail? or even just call `conn->options()` on the returned connection?

**Review by coryan — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10748)* status: 0 of 15 files reviewed, 2 unresolved discussions (waiting on @dbolduc)

___
*[`generator/integration_tests/golden/v1/golden_thing_admin_rest_connection.h` line 57 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10748#-NNOHoWR2LMbtlpONDEY-r1-57:-NNO_UUj6sWt_cjM8j51:b-79kfvt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63d1409478890e431e3aee06fccf8057c08d4afa/generator/integration_tests/golden/v1/golden_thing_admin_rest_connection.h#L57)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Can we have a unit test for this thing?

I know these are kinda weird to test... but can we do something like set a fake local endpoint, and make one client call we know will fail? or even just call `conn->options()` on the returned connection?
</blockquote></details>


Test added.
___
*[`generator/internal/descriptor_utils.cc` line 845 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10748#-NNOHoWR2LMbtlpONDEe-r1-845:-NNOaRMu8Zxrdqpj_dUR:brgo0ya) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63d1409478890e431e3aee06fccf8057c08d4afa/generator/internal/descriptor_utils.cc#L845)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

are we concerned that a customer might see the rest connection header, but have no idea how to enable it?

(unlikely, as we know that Spanner Admin APIs are not used.... but this problem seems avoidable)
</blockquote></details>

At this point in time, we don't want them to enable it. But when we decide to widely roll out transcoding, we will need some documentation.

**Review by dbolduc — APPROVED** [review]

File: generator/internal/descriptor_utils.cc (hunk 2 above)
Comment (scotthart) [reply generator/internal/descriptor_utils.cc:840]: Added.

File: generator/internal/connection_rest_generator.cc (hunk 3 above)
Comment (scotthart) [reply generator/internal/connection_rest_generator.cc]: Removed.

File: generator/internal/connection_rest_generator.cc (hunk 4 above)
Comment (scotthart) [reply generator/internal/connection_rest_generator.cc]: Done.

File: generator/internal/connection_rest_generator.cc (hunk 5 above)
Comment (scotthart) [reply generator/internal/connection_rest_generator.cc]: Done.

File: generator/internal/connection_rest_generator.cc [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+Status ConnectionRestGenerator::GenerateHeader() {
+  HeaderPrint(CopyrightLicenseFileHeader());
+  HeaderPrint(R"""(
+// Generated by the Codegen C++ plugin.
+// If you make any local changes, they will be lost.
+// source: $proto_file_name$
+
+#ifndef $header_include_guard$
+#define $header_include_guard$
+
```
Comment (devbww) [inline generator/internal/connection_rest_generator.cc:46]: One too many blank lines?

File: generator/internal/connection_rest_generator.cc [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+Status ConnectionRestGenerator::GenerateCc() {
+  CcPrint(CopyrightLicenseFileHeader());
+  CcPrint(R"""(
+// Generated by the Codegen C++ plugin.
+// If you make any local changes, they will be lost.
+// source: $proto_file_name$
+
+#ifndef $header_include_guard$
+#define $header_include_guard$
+
```
Comment (devbww) [inline generator/internal/connection_rest_generator.cc:138]: One too many blank lines.

File: generator/internal/connection_rest_generator.cc (hunk 8 above)
Comment (scotthart) [reply generator/internal/connection_rest_generator.cc]: Fixed.

File: generator/internal/connection_rest_generator.cc (hunk 9 above)
Comment (scotthart) [reply generator/internal/connection_rest_generator.cc]: Fixed.

**Pull request merged**